### PR TITLE
Enable strict null checks

### PR DESCRIPTION
**Pull request by zharinov**
- ref renovatebot/renovate#12869

### CODE REVIEW

File: .eslintignore [hunk 1]
```diff
@@ -14,3 +14,6 @@ coverage
 **/*.generated.ts
 /tools/dist
 /patches
+
+.eslintrc.js
+jest.config.ts
```
Comment (viceice) [inline .eslintignore:19]: Nooooo. Why?

File: lib/index.ts [hunk 2]
```diff
@@ -13,52 +13,57 @@ export type ReleaseType =
 /**
  * v1 == v2 This is true if they're logically equivalent, even if they're not the exact same string. You already know how to compare strings.
  */
-export function eq(v1: string, v2: string): boolean {
+export function eq(v1: string, v2: string): boolean | null {
   const x = Version.create(v1);
   const y = Version.create(v2);
+  if (x === null || y === null) return null;
```
Comment (viceice) [inline lib/index.ts:19]: Do we really want to have `null` ? Shouldn't we simply return false?

File: tsconfig.lint.json [hunk 3]
```diff
@@ -1,4 +0,0 @@
-{
-  "extends": "./tsconfig",
-  "include": [".eslintrc.js", "**/*.ts", "**/*.js", "**/*.mjs"]
-}
```
Comment (viceice) [inline tsconfig.lint.json:4]: Please use this for lint and include all js files. We can disable strict flags for lint if required

File: package.json [hunk 4]
```diff
@@ -29,23 +29,26 @@
...
+    "eslint-plugin-only-warn": "1.0.3",
+    "jest": "27.4.5",
+    "jest-junit": "13.0.0",
     "jest-silent-reporter": "0.5.0",
     "npm-run-all": "4.1.5",
     "prettier": "2.5.1",
-    "ts-jest": "26.5.6",
+    "ts-jest": "27.1.2",
+    "ts-node": "10.4.0",
+    "ttypescript": "1.5.13",
```
Comment (viceice) [inline package.json:51]: What is `ttypescript` used for?

File: lib/ruby/requirement.ts [hunk 5]
```diff
@@ -16,30 +16,37 @@
 
 import { Version } from './version';
```
Comment (viceice) [inline lib/ruby/requirement.ts:17]: Maybe a type import only ?
```suggestion
import type { Version } from './version';
```

File: .eslintrc.js [hunk 6]
```diff
@@ -1,71 +1,61 @@
...
-    '@typescript-eslint/no-var-requires': 'off', // disable until all files converted to typescript
-    '@typescript-eslint/no-use-before-define': 'off', // disable until all files converted to typescript
-    '@typescript-eslint/explicit-member-accessibility': 0,
-    '@typescript-eslint/explicit-function-return-type': 0,
-    '@typescript-eslint/no-explicit-any': 0,
-    '@typescript-eslint/no-non-null-assertion': 0,
+    // Same approach used by TypeScript noUnusedLocals
     '@typescript-eslint/no-unused-vars': [
-      'error',
+      'warn',
```
Comment (viceice) [inline .eslintrc.js:24]: why no error?

File: lib/ruby/requirement.ts (hunk 5 above)
Comment (zharinov) [reply lib/ruby/requirement.ts:17]: Won't work as it's a constructor

File: lib/index.ts [hunk 7]
```diff
@@ -83,7 +88,7 @@ export function maxSatisfying(
   versions: string[],
   range: string
 ): string | null {
-  return versions.reduce((x, y) => {
+  return versions.reduce<string | null>((x, y) => {
```
Comment (viceice) [inline lib/index.ts:91]: why `null`? how can it happen?

File: lib/ruby/requirement.ts [hunk 8]
```diff
@@ -130,14 +137,14 @@ export class Requirement {
   //       end
   //     end
   //   end
-  static create(...inputs: any[]): Requirement {
+  static create(...inputs: unknown[]): Requirement {
```
Comment (viceice) [inline lib/ruby/requirement.ts:140]: Can we be more type safe here?

File: lib/ruby/requirement.ts [hunk 9]
```diff
@@ -290,7 +287,8 @@ export class Requirement {
   //   end
   isNone(): boolean {
     if (this._requirements.length === 1) {
-      const [op, v] = this._requirements[0];
+      const req: ParsedRequirement = this._requirements[0] as never;
```
Comment (viceice) [inline lib/ruby/requirement.ts:290]: Even more bad

File: tsconfig.build.json [hunk 10]
```diff
@@ -0,0 +1,4 @@
+{
+  "extends": "./tsconfig.json",
+  "exclude": ["node_modules", "lib/**/*.spec.ts"]
```
Comment (viceice) [inline tsconfig.build.json:3]: We should make other way around and only include our source files

File: lib/index.ts (hunk 7 above)
Comment (zharinov) [reply lib/index.ts:91]: It's null when neither of the versions satisfies the range

**Comment by zharinov** [comment]
Well, I didn't expect it would require such a massive rework.
The code was meant to be as close to the original Ruby code as possible, but null safety means we need to re-design the API get rid of the most "safety nets".
I'm stepping back for now, as this task seems to be not important enough. Sorry for that.

@viceice Thanks for review 👍

**Pull request closed without merging**